Adding Hdf5 Functionality

P-1884 opened the pull request:
@JelleAalbers @swagnercarena As mentioned earlier, here's the code to save the images as compressed hdf5 files rather than numpy. This is added as a kwarg --h5 when running generate.py and train_model.py.

JelleAalbers commented on paltas/Analysis/dataset_generation.py:
```diff
@@ -151,7 +151,7 @@ def kwargs_detector_to_tf_noise(image,kwargs_detector):
 
 def generate_tf_record(npy_folder,learning_params,metadata_path,
-	tf_record_path):
+	tf_record_path,h5=False):
```
Could you add the new argument to the docstring?

JelleAalbers commented on paltas/Analysis/dataset_generation.py:
```diff
@@ -15,7 +15,7 @@
 import warnings
 from scipy.ndimage import rotate
-
+import h5py
```
Is h5py already installed by one of our dependencies? If not, you could do e.g.
```python
try:
    import h5py
except ImportError:
    h5py = None
```
and maybe print a useful message if someone tries to use  `--h5` without having h5py installed. Or we could just add h5py to our requirements.txt -- if you end up writing a unit test for this, that would be the easiest move.

JelleAalbers commented on paltas/generate.py:
```diff
@@ -41,6 +41,8 @@ def parse_args():
 	parser.add_argument('--tf_record', action='store_true',
 		help='Generate the tf record for the training set.')
+	parser.add_argument('--h5', action='store_true',
+		help='Save images as .h5 files rather than .npy')
```
You actually save all the images to a single h5 file named image_data.h5

JelleAalbers reviewed (commented):
Thanks Phil, this looks good; I think we'll certainly want to have this feature. 
  * I commented on a few small code style conventions. We kind of follow the guidelines that lenstronomy has (https://lenstronomy.readthedocs.io/en/latest/contributing.html#general-guidelines), i.e. PEP8 with a few exceptions and not being super strict about everything.
  * If you'd like to try writing a unit test, you could add something to the tests here https://github.com/swagnercarena/paltas/blob/main/test/generate_tests.py to check that images are actually being written to the h5 files.
  * In the future we might think about writing the metadata and the config to the hdf5 as well, so that everything is contained in one file. It shouldn't increase the image data much. But that's a new feature request, shouldn't hold up your PR I think.

drphilmarshall commented on paltas/Analysis/dataset_generation.py:
```diff
@@ -177,16 +179,31 @@ def generate_tf_record(npy_folder,learning_params,metadata_path,
+				image_shape = image_i.shape
+			else:
+				npy_file = npy_file_list[file_number]
 			# Pull the index from the filename
```
PEP8 style is indent with just 4 spaces, too.

P-1884 commented:
Hi Jelle, Thanks for these! I've gone through each one and edited the corresponding sections. I've added a unit test as well - this is very similar to the test_main function in generate_tests.py but for the h5 rather than npy files. I've also added h5py to the requirements list as well.

JelleAalbers commented on paltas/generate.py:
```diff
@@ -104,8 +107,30 @@ def main():
 				header=first_write)
 			metadata_list = []
-
+#
```
Stray # (also one a few lines below)

JelleAalbers commented on test/generate_tests.py:
```diff
@@ -156,3 +156,69 @@ def test_main_drizzle(self):
+		generate.main()
+
+		image_file_list = glob.glob(os.path.join(output_folder,'image_data.h5'))
+		# Check only 1 h5 file is made:
```
There cannot be two files with the same name, so you're really just checking the file exists here. Maybe change the glob pattern to *.h5?

JelleAalbers commented on test/generate_tests.py:
```diff
@@ -156,3 +156,69 @@ def test_main_drizzle(self):
+			self.assertTupleEqual(img.shape,(64,64))
+		os.remove(image_file_list[0])
+
+		# Make sure the metadata makes sense
```
This block seems copied verbatim from the above tests, so could be factored out into a function. But we can also fix that later, this is  not the only place in our test suite with some duplication..

JelleAalbers commented on test/generate_tests.py:
```diff
@@ -156,3 +156,69 @@ def test_main_drizzle(self):
+
+		# Also clean up the test cosmos cache
+		test_cosmo_folder = 'test_data/cosmos/'
+		
```
You'll also want to remove the h5 file here; currently the test fails when it tries to remove the non-empty directory in the final rmdir call.

JelleAalbers commented on test_requirements.txt:
```diff
@@ -1 +1,2 @@
-tensorflow>=2.4.1
\ No newline at end of file
+tensorflow>=2.4.1
+h5py
```
You'll want to add this to requirements.txt if you do a plain "import h5py" (without the try: except ImportError I suggested above) in the main package, because then paltas won't run if it is not installed.

JelleAalbers commented on test/generate_tests.py:
```diff
@@ -156,3 +156,69 @@ def test_main_drizzle(self):
+		image_file_list = glob.glob(os.path.join(output_folder,'image_data.h5'))
+		# Check only 1 h5 file is made:
+		self.assertEqual(len(image_file_list),1)
+		with h5py.File(image_file_list[0],'r') as f0:
```
The test fails here currently because there is no import h5py at the top of the file

JelleAalbers reviewed (commented):
Thanks Phil! Just a few small comments mostly to get the tests to pass, then I think this is ready to go.

JelleAalbers commented on test/generate_tests.py:
```diff
@@ -156,3 +156,69 @@ def test_main_drizzle(self):
+	def test_h5(self):
+		# Test that the h5 kwarg generates the required h5 files
+		old_sys = copy.deepcopy(sys.argv)
+		n_generate = 21
```
I think you're currently not testing the 100-image batching logic; maybe try generating 101 instead of 21 images? The tests will take a bit longer, but we can always optimize the test config to be a simpler task later.

P-1884 commented:
Hi Jelle, Thanks for these! I've fixed what you suggested - the biggest change is I've factored out the metadata checks in the generate_tests file to remove duplication. These checks now pass when I'm running it locally so should (hopefully!) still pass on Github too.

JelleAalbers reviewed (approved):
Thanks for the nice addition Phil and for addressing all the comments, let's merge this!

(Final nitpick is that h5py doesn't have to be in test_requirements.txt anymore if it's already in requirements.txt, but we can fix that afterwards, it doesn't hurt either.)

swagnercarena reviewed (changes requested):
Everything here looks great, I only request a renaming of the variable n_npy_files. It no longer represents the number of npy files but rather the number of images. I think keeping the current naming reduces readability. Once that's changed, I'm happy to merge.

P-1884 commented:
Thanks, should be all sorted now! I've also removed the h5py requirement from the text_requirements doc :)

swagnercarena commented:
@P-1884 go ahead and hit merge whenever you want.

P-1884 commented:
Thanks Sebastian! I don't think I have write access to paltas yet - could you merge it (or alternatively how can I get write-access?)
